h2o-py/h2o/backend/cluster.py

```diff
@@ -29,42 +31,30 @@ class H2OCluster(object):
     def __init__(self):
         """Initialize new H2OCluster instance."""
         self._props = {}
-        self._retrieved_at = None
+        self._retrieved_at = time.time()
 
-    @staticmethod
-    def from_kvs(keyvals):
+    @classmethod
+    def make(cls, keyvals):
         """
         Create H2OCluster object from a list of key-value pairs.
-
-        TODO: This method should be moved into the base H2OResponse class.
         """
-        obj = H2OCluster()
-        obj._retrieved_at = time.time()
+        obj = cls()
         for k, v in keyvals:
-            if k in {"__meta", "_exclude_fields", "__schema"}: continue
-            if k in _cloud_v3_valid_keys:
-                obj._props[k] = v
-            else:
-                raise AttributeError("Attribute %s cannot be set on H2OCluster (= %r)" % (k, v))
+            if k in _ignored_schema_keys: continue
+            if not hasattr(cls, k):
+                # we can add properties dynamically here as they are defined statically on the backend
+                # as properties of water.api.schemas3.CloudV3
+                setattr(cls, k, property(partial(cls.__getitem__, name=k)))
```

Comment:
to avoid having to raise an unnecessary error when there is a version mismatch between PY client and backend, the properties are now generated on the fly on the first call, adapting dynamically to the backend schema.
This also avoid developers to update this code when the schema is modified.

Reply:
@michalkurka as it is potentially the most impacting change in this PR: do you validate this?

Reply:
@sebhrusen I believe the intention here was the exact list on the client so that there is a clear understanding of what each returned property means.

My preference would be to keep the list static and well-defined. IMHO this is a bad property to have "This also avoid developers to update this code when the schema is modified.". If something is added to the cloud response it should definitely fail - why is it added if it is not being used? should it be added in the first place? I think the list of the properties establishes a "protocol" that the client and cluster can rely on.

I think we should definitely fail if expected properties are not found. If there are some additional properties we can consider allowing user to override using the strict_version_check. To override missing properties I think users would need to switch to "god mode".

My opinion :)

Reply:
My point is that the version check should be done when connecting to the backend, so any additional error at runtime because the cluster schema on the backend doesn't match the cluster schema on the frontend is just an hindrance: user may still want to use `strict_version_check=True` but in practice it will still fail anyway on the first 'ping'.

So even if the static API for this particular component was generated the same way as for estimators, we would still have this issue. That's why, by generating the properties on the fly, I'm just suggesting to use the REST schema as the single point of reference: I could also make a request to `/3/Metadata/schemas/CloudV3` (or sth like that) to generate the properties from there if you think it's cleaner (I think it is, but the result is equivalent): that's what we do in Flow for example.

Reply:
> If something is added to the cloud response it should definitely fail - why is it added if it is not being used? should it be added in the first place? I think the list of the properties establishes a "protocol" that the client and cluster can rely on.

@michalkurka I think it makes sense only if you put the client at the same level as the backend. But I consider the REST API as the only point of reference / contract that both backend and clients should respect.
Given this, if something is available in the schema, the client should adapt.

Reply:
> strict_version_check=True but in practice it will still fail anyway on the first 'ping'.

it shouldn't fail on with a non-descript message, also I think having an override to allow new unrecognized properties is fine as well

> I think it makes sense only if you put the client at the same level as the backend
yes, exactly my thinking - the client and backend are in normal circumstances "one entity" - if someone wants to use strict_version_check to debug something and mix it  - fine with me but it should at least alert them there is a mismatch on the fundamental information that the client and backend exchanges

Flow is not a client in the same sense as the others are - Flow is inherently tied to the same H2O version, it doesn't exist beyond the same version actually; it is just dynamically doing what eg. "gen_R.py" is doing at compile time

Again - this is my preference - I am not going to force you to do it that way.

Reply:
> Again - this is my preference - I am not going to force you to do it that way.

I understand, but I need your always pertinent ;) arguments before deciding what I want to do.

Reply:
FYI, I think I'll keep it dynamic, but will extract the properties directly from the `Metadata/schema` API, and also address your concern regarding the error message when client tries to access unknown properties: I agree that in this case the error message should be descriptive and inform that he's trying to use a property that is not available for the specific backend to which it is connected.